system/settings/storage_bin.c

```diff
@@ -0,0 +1,246 @@
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: load_bin
+ *
+ * Description:
+ *    Loads binary data from a storage file.
+ *
+ * Input Parameters:
+ *    file             - the filename of the storage to use
+ *
+ * Returned Value:
+ *   Success or negated failure code
+ *
+ ****************************************************************************/
+
+int load_bin(FAR char *file)
+{
+  static int fd;
+  int i;
+  int ret = OK;
+  uint16_t valid;
+  uint16_t count;
+
+  fd = open(file, O_RDONLY);
+  if (fd < 0)
+    {
+      return -ENOENT;
+    }
+
+  valid = 0;
+  read(fd, &valid, sizeof(uint16_t));
+
+  if (valid != VALID)
+    {
+      goto abort; /* Just exit - the settings aren't valid. What to do? */
+    }
+
+  count = 0;
+  read(fd, &count, sizeof(uint16_t));
+
+  uint32_t calc_crc = 0;
+  uint32_t exp_crc = 0;
+
+  for (i = 0; i < count; i++)
+    {
+      setting_t setting;
+      read(fd, &setting, sizeof(setting_t));
+      calc_crc = crc32part((FAR uint8_t *)&setting, sizeof(setting_t),
+                            calc_crc);
+    }
+
+  read(fd, &exp_crc, sizeof(uint32_t));
+
+  if (calc_crc != exp_crc)
+    {
+      ret = -EIO;
+      goto abort;
+    }
+
+  lseek(fd, (sizeof(uint16_t) *2), SEEK_SET);  /* Get after valid & size */
+
+  for (i = 0; i < count; i++)
+    {
+      setting_t setting;
+      read(fd, &setting, sizeof(setting_t));
+
+      FAR setting_t *slot = getsetting(setting.key);
+      if (slot == NULL)
+        {
+          continue;
+        }
+
+      memcpy(slot, &setting, sizeof(setting_t));
+    }
+
+abort:
+  close(fd);
+  return ret;
+}
+
+/****************************************************************************
+ * Name: save_bin
+ *
+ * Description:
+ *    Saves binary data to a storage file.
+ *
+ * Input Parameters:
+ *    file             - the filename of the storage to use
+ *
+ * Returned Value:
+ *   Success or negated failure code
+ *
+ ****************************************************************************/
+
+int save_bin(FAR char *file)
+{
+  FAR uint8_t *buffer = malloc(BUFFER_SIZE);
+  int count;
+  int fd;
+  int ret = OK;
+
+  if (buffer == NULL)
+    {
+      return -ENOMEM;
+    }
+
+  count = 0;
+  int i;
+  for (i = 0; i < CONFIG_SETTINGS_MAP_SIZE; i++)
+    {
+      if (g_settings.map[i].type == SETTING_EMPTY)
+        {
+          break;
+        }
+
+      count++;
+    }
+
+  fd = open(file, (O_RDWR | O_TRUNC), 0666);
+  if (fd < 0)
+    {
+      ret = -ENODEV;
+      goto abort;
+    }
+
+  memset(buffer, 0xff, BUFFER_SIZE);
+  *((FAR uint16_t *)buffer) = VALID;
+  *(((FAR uint16_t *)buffer) + 1) = count;
+
+  off_t offset = sizeof(uint16_t) *2;  /* Valid & count */
```

Comment:
```suggestion
  off_t offset = sizeof(uint16_t) * 2;  /* Valid & count */
```
move the declaration to the beginning of function, same for data_size, rem_data, crc and rem_crc.